### PR TITLE
Make inventory hierarchy dynamic

### LINKED ISSUE #93: Make the Inventory Structure Configurable
# Background
Currently the inventory structure is hard coded into Commodore:
* `global.common`
* `global.cloud.<cloud-provider>`
* `global.cloud.<cloud-provider>.<cloud-region>`
* `global.distribution.<k8s-distribution>`
* `<customer>.<cluster>`

This limits the flexibility of different setups and makes changes to this hierarchy more difficult.

Similar to Puppet Hiera, this structure should be configurable outside of Commodore.

# Proposal
Implement a config file which can define the inventory structure. This config file should be created in the global config Git repo with a specific name (e.g. `syn.yml`). It should allow to use facts as placeholders for names in the structure. The leaf of an entry is always a class (i.e. `.yml` file).

The referenced classes should be optional and if they don't exist they should be skipped. I.e. if a cloud doesn't have a region, no `cloud/${cloud}/${region}.yml` file needs to exist.

Possible example:
```yaml
hierarchy:
  - global/common
  - global/cloud/${cloud}
  - global/cloud/${cloud}/${region}
  - global/distribution/${distribution}
  - global/lieutenant-instance/${lieutenant-instance}
  - ${tenant}/common
  - ${tenant}/${cluster}
```
# References
* Puppet Hiera config: https://puppet.com/docs/hiera/3.3/configuring.html

### PR DESCRIPTION
**Pull request by corvus-ch**
Resolves #93 

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Update tests.
- [x] Update the ./CHANGELOG.md.
- [x] Link this PR to related issues.

### CODE REVIEW

**Comment by corvus-ch** [comment]
This is currently being blocked by https://github.com/deepmind/kapitan/pull/610.

**Comment by corvus-ch** [comment]
I am currently writing the documentation. @simu @srueg could you please already review the code changes? Please note the two commits starting with `TEMP` in the message. Those two commits will be removed before this PR can be merged. They are workarounds for changes to be done elsewhere.

File: commodore/cluster.py
Comment (srueg) [inline commodore/cluster.py:87]: Is this still something we should enforce?
I'm just thinking how generic these really are. We don't have any dependency on hard coded facts anymore anywhere else...

File: commodore/cluster.py
Comment (srueg) [inline commodore/cluster.py:120]:
```suggestion
    return P('inventory', 'classes', 'params') / f"{target}.yml"
```

File: commodore/compile.py
Comment (srueg) [inline commodore/compile.py:96]: Since it looks like we only need the cluster and tenant IDs anymore, can we change `reconstruct_api_response()` to only get these two values?

**Review by srueg — CHANGES_REQUESTED** [review]

File: commodore/cluster.py
Comment (corvus-ch) [reply commodore/cluster.py:50]: I am not sure. Maybe @simu does have an opinion on that.

File: commodore/cluster.py
Comment (corvus-ch) [reply commodore/cluster.py:87]: You are right. This is no longer needed. I will remove it.

Comment (corvus-ch) [reply commodore/cluster.py:87]: Um wait! We need to keep it until we remove the deprecated `cluster.dist` and `cloud.provider` parameters.

File: commodore/compile.py
Comment (corvus-ch) [reply commodore/compile.py]: Refactored to `read_cluster_and_tenant(target: str) -> Tuple[str, str]`.

File: pyproject.toml
Comment (srueg) [inline pyproject.toml:16]:
```suggestion
kapitan = { git = "https://github.com/deepmind/kapitan.git", rev = "bfeda63836e5fbd220f3e4cd4bd490d9e1a7eb09" }
```

**Comment by corvus-ch** [comment]
Turns out deepmind/kapitan#610 was not needed after all. I removed the Kapitan update as it was braking the build.

File: commodore/cluster.py
Comment (tobru) [reply commodore/cluster.py:44]: I would say: yes.

File: docs/modules/ROOT/pages/reference/hierarchy.adoc
Comment (tobru) [inline docs/modules/ROOT/pages/reference/hierarchy.adoc:1]:
```suggestion
= Configure the Inventory Hierarchy
```

File: docs/modules/ROOT/pages/reference/hierarchy.adoc
Comment (tobru) [inline docs/modules/ROOT/pages/reference/hierarchy.adoc:3]: Better link to the fork? https://github.com/kapicorp/reclass

File: docs/modules/ROOT/pages/reference/concepts.adoc
Comment (srueg) [inline docs/modules/ROOT/pages/reference/concepts.adoc:18]:
```suggestion
The hierarchy could then be extended by the following levels.
```

File: docs/modules/ROOT/pages/reference/hierarchy.adoc
Comment (srueg) [inline docs/modules/ROOT/pages/reference/hierarchy.adoc:37]:
```suggestion
To build up your configuration hierarchy, include classes in `common.yml` within your global configuration inventory.
```

File: docs/modules/ROOT/pages/reference/hierarchy.adoc
Comment (srueg) [inline docs/modules/ROOT/pages/reference/hierarchy.adoc:65]:
```suggestion
When building a hierarchy that has optional parts, one must be creative.
```

**Review by tobru — APPROVED** [review]
Two small nitpicks - otherwise LGTM. Let's see how we continue with the documentation, I think the whole hierarchy stuff needs more. But for the time being that's good enough.

File: docs/modules/ROOT/partials/nav-reference.adoc
Comment (srueg) [inline docs/modules/ROOT/partials/nav-reference.adoc:5]:
```suggestion
* xref:commodore:ROOT:reference/hierarchy.adoc[Inventory Hierarchy]
```

**Review by srueg — CHANGES_REQUESTED** [review]

File: commodore/cluster.py
Comment (simu) [reply commodore/cluster.py:44]: Commodore used to print a warning about 'old-style components' when a component did not have a `defaults.yml`, but it appears that has been removed at some point.

File: docs/modules/ROOT/pages/reference/concepts.adoc
Comment (simu) [inline docs/modules/ROOT/pages/reference/concepts.adoc:15]: With the change the whole structure of this section is pretty confusing, as the configuration value sources can now arbitrarily be changed.

I think the part of the section which starts with "Commodore defines the following hierarchy" needs to be completely replaced with a description of the new dynamic inventory mechanism, and a description of how component defaults fit into the new picture.

File: docs/modules/ROOT/pages/reference/hierarchy.adoc
Comment (simu) [reply docs/modules/ROOT/pages/reference/hierarchy.adoc:1]: The Microsoft style guide (which we're using according to our documentation) says to use sentence case for titles (cf. https://docs.microsoft.com/en-us/style-guide/capitalization#sentence-style-capitalization-in-titles-and-headings)

File: docs/modules/ROOT/pages/reference/hierarchy.adoc
Comment (simu) [inline docs/modules/ROOT/pages/reference/hierarchy.adoc:4]: Note: Upstream Reclass doesn't allow parameters in classes. That's one of the features of Kapitan's fork: https://github.com/kapicorp/reclass/blob/develop/README-extensions.rst#use-references-in-class-names

**Review by simu — COMMENTED** [review]
LGTM overall. Some comments regarding the docs.

File: docs/modules/ROOT/pages/reference/hierarchy.adoc
Comment (srueg) [reply docs/modules/ROOT/pages/reference/hierarchy.adoc:4]: Note: It's not really the Kapicorp fork, but the salt-formula fork: https://github.com/salt-formulas/reclass
(which in turn is a fork of the original madduck reclass: https://github.com/madduck/reclass)

File: docs/modules/ROOT/pages/reference/hierarchy.adoc
Comment (corvus-ch) [reply docs/modules/ROOT/pages/reference/hierarchy.adoc:1]: The vale checker does not assert this but the Microsoft style guide asks for sentence case in titles.

File: commodore/cluster.py
Comment (corvus-ch) [reply commodore/cluster.py:44]: This is already covered. See https://github.com/projectsyn/commodore/blob/159b294627313cf54a237e43d835e493644dcd8b/commodore/dependency_mgmt.py#L30 and https://github.com/projectsyn/commodore/blob/159b294627313cf54a237e43d835e493644dcd8b/commodore/component/compile.py#L129.

File: docs/modules/ROOT/pages/reference/concepts.adoc
Comment (corvus-ch) [reply docs/modules/ROOT/pages/reference/concepts.adoc]: I have rewritten that part.

File: docs/modules/ROOT/pages/reference/hierarchy.adoc
Comment (corvus-ch) [reply docs/modules/ROOT/pages/reference/hierarchy.adoc]: I have now explicitly stated, that a fork is being used. Including link to the extension documentation of that fork.

**Review by simu — APPROVED** [review]

**Review by srueg — APPROVED** [review]

File: commodore/cluster.py
Comment (srueg) [inline commodore/cluster.py:75]:
```suggestion
            'global.lieutenant-instance.${facts:lieutenant-instance}',
```

Comment (srueg) [reply commodore/cluster.py:75]: For some reason the `lieutenant-instance` class is not properly included if I try to compile the `c-old-fire-5788` cluster.

File: commodore/cluster.py
Comment (srueg) [inline commodore/cluster.py:48]: This needs to be included after `global.common`

File: commodore/cluster.py
Comment (srueg) [reply commodore/cluster.py:75]: Hmm, seems like the referencing breaks because the order of import changed: As mentioned above, `global.common` needs to be included before `common` for it to work properly.

Comment (srueg) [reply commodore/cluster.py:75]: I did some tests and I fear we have to split the hierarchy definition and parameters into two files.
Currently both is defined in [common.yml](https://git.vshn.net/syn/commodore-defaults/-/blob/dynamic_hierarchy/common.yml). But this has the effect, that references used in there (e.g. `VAULT_ADDR: ${secret_management:vault_addr}`) are not properly overwritten, because they are evaluated first, before including other classes.

We could for example leave only the class imports in `common.yml` and move all the `parameters` definitions into a new file (e.g. `global.params`) which needs to be included as the first class.

**Review by srueg — CHANGES_REQUESTED** [review]

File: commodore/cluster.py
Comment (corvus-ch) [reply commodore/cluster.py:48]: Resolved by splitting class includes and parameters into different files.

**Review by simu — APPROVED** [review]

**Review by srueg — APPROVED** [review]

**Pull request merged**
